appveyor.yml

```diff
@@ -1,41 +1,57 @@
-image: Visual Studio 2017
+image:
+- Ubuntu
+- Visual Studio 2017
 
 skip_tags: false
 
 test: off
 
-init:
-- ps: >-
-    if ($env:APPVEYOR_REPO_TAG -eq "true")
-    {
-        Update-AppveyorBuild -Version ($env:APPVEYOR_REPO_TAG_NAME).TrimStart("v")
-    }
-
-after_build:
-- ps: >-
-    $wc = New-Object System.Net.WebClient
-
-    $wc.UploadFile("https://ci.appveyor.com/api/testresults/nunit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path "NetCoreClrTestResults.xml"))
-
-    $wc.UploadFile("https://ci.appveyor.com/api/testresults/nunit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path "NetCoreClrWeakNamedTestResults.xml"))
-
-    $wc.UploadFile("https://ci.appveyor.com/api/testresults/nunit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path "DesktopClrTestResults.xml"))
-
-    $wc.UploadFile("https://ci.appveyor.com/api/testresults/nunit/$($env:APPVEYOR_JOB_ID)", (Resolve-Path "DesktopClrWeakNamedTestResults.xml"))
-
-on_success:
-- ps: >-
-    if ($env:APPVEYOR_REPO_TAG -eq "true")
-    {
-        nuget push ".\build\Castle.Core.${env:APPVEYOR_BUILD_VERSION}.nupkg" -ApiKey $env:NUGET_API_KEY -Source https://api.nuget.org/v3/index.json
-        nuget push ".\build\Castle.Core-log4net.${env:APPVEYOR_BUILD_VERSION}.nupkg" -ApiKey $env:NUGET_API_KEY -Source https://api.nuget.org/v3/index.json
-        nuget push ".\build\Castle.Core-NLog.${env:APPVEYOR_BUILD_VERSION}.nupkg" -ApiKey $env:NUGET_API_KEY -Source https://api.nuget.org/v3/index.json
-        nuget push ".\build\Castle.Core-Serilog.${env:APPVEYOR_BUILD_VERSION}.nupkg" -ApiKey $env:NUGET_API_KEY -Source https://api.nuget.org/v3/index.json
-    }
-
-build_script:
-- cmd: build.cmd
-
-artifacts:
-- path: build\*.nupkg
-  name: core
+for:
+-
+  matrix:
+    only:
+    - image: Ubuntu
+
+  build_script:
+  - uname -a
+  - mono --version
+  - ./build.sh
+
+  deploy: off
+
+-
+  matrix:
+    only:
+    - image: Visual Studio 2017
+
+  init:
```

Comment:
I'm having a hard time reading the YAML here to work out which sections belong to Windows and which apply to both. Can you add some block comments or something to make this clearer, I wanted to make sure only one job publishes to nuget.org, but both should upload test results and artifacts.

Can we do anything to make sure that we only publish to nuget.org if both builds succeed? Is that one of the future things you want to do?

Reply:
> Can you add some block comments or something to make this clearer

Will do.

> Can we do anything to make sure that we only publish to nuget.org if both builds succeed? 

I *think* that's what will already happen, but it's a little difficult to verify.

Right now, because we're on the Free plan, we only get 1 concurrent job, so one build's jobs run sequentially. If we put the Linux job first and additionally enable fail-fast (`matrix.fast_finish: true`) then the Windows build will never run and deployment cannot happen.

The above strikes me as inelegant, we still want results from the Windows build even if the Linux build job failed. So I'd rather have a guarantee from AppVeyor that any `on_success` script or `deploy` section only runs if all build jobs have been successful, even when the `on_success` script is specific to one job & another job is still running in parallel.

I'm still checking with the AppVeyor docs, I'll report back my findings.

> both should upload test results and artifacts

Yes, that would be a good addition. I'll add the upload for the Linux build.

Reply:
> Can we do anything to make sure that we only publish to nuget.org if both builds succeed?

Not easily. From the [AppVeyor Discussions forum](https://help.appveyor.com/discussions/questions/1164-how-do-i-run-a-script-after-all-matrix-possibilities-succeed):

> All jobs run independently and `on_success` is part of the job, not a build

What we have now&mdash;two jobs that execute sequentially, letting the last one handle deployment&mdash;combined with the fail-fast strategy mentioned above would be the easiest option.

We could, of course, look at other CI offers such as Microsoft's Azure Pipelines to see if they offer the flexibility we're looking for...

Reply:
Let's just let them continue to run in whatever order and only publish in the Windows build, it will be no different to what we have now with AppVeyor and Travis being completely separate.

It shouldn't be a problem anyway because pull requests build, then master builds, then finally you tag and get another build so should have picked up by then if the Linux build will fail.

Reply:
Done. I added some comments to `appveyor.yml` to make it easier to read, and I restored the Windows build to run first. Regarding the test counts stated in AppVeyor's UI, these don't add up from the test count reported in the log, I dare say that's more of a problem with the AppVeyor UI than with the way how we upload test results from `appveyor.yml`.